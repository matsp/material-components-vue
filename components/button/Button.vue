<template>
  <a
    v-if="href"
    :class="classes"
    :href="href"
    v-bind="$attrs"
    role="button"
    class="mdc-button"
    v-on="$listeners"
  >
    <slot name="icon" />
    <slot />
  </a>
  <button
    v-else
    :class="classes"
    v-bind="$attrs"
    class="mdc-button"
    v-on="$listeners"
  >
    <slot name="icon" />
    <slot />
  </button>
</template>

<script>
import { MDCRipple } from '@material/ripple'
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    raised: {
      type: Boolean,
      default: false
    },
    unelevated: {
      type: Boolean,
      default: false
    },
    outlined: {
      type: Boolean,
      default: false
    },
    dense: {
      type: Boolean,
      default: false
    },
    href: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      mdcRipple: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-button--raised': this.raised,
        'mdc-button--unelevated': this.unelevated,
        'mdc-button--outlined': this.outlined,
        'mdc-button--dense': this.dense
      }
    }
  },
  watch: {
    classes () {
      this.mdcRipple.destroy()
      this.mdcRipple = MDCRipple.attachTo(this.$el)
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    this.mdcRipple = MDCRipple.attachTo(this.$el)
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
  },
  methods: {
    updateSlot () {
      if (this.$slots.icon) {
        this.$slots.icon.map(n => {
          n.elm.classList.add('mdc-button__icon')
          n.elm.setAttribute('aria-hidden', 'true')
        })
      }
    }
  }
}
</script>
