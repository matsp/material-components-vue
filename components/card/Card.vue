<template>
  <div
    :class="classes"
    class="mdc-card"
  >
    <div
      v-if="$slots['actionableContent']"
      ref="content"
      :class="contentClasses"
      tabindex="0"
    >
      <slot name="actionableContent" />
    </div>
    <slot />
    <div
      v-if="$slots['actionButtons'] || $slots['actionIcons'] || $slots['fullBleedButton']"
      :class="actionClasses"
      class="mdc-card__actions"
    >
      <slot
        v-if="fullBleedAction"
        name="fullBleedButton"
      />
      <div
        v-if="!fullBleedAction && $slots['actionButtons']"
        class="mdc-card__action-buttons"
      >
        <slot name="actionButtons" />
      </div>
      <div
        v-if="!fullBleedAction && $slots['actionIcons']"
        class="mdc-card__action-icons"
      >
        <slot name="actionIcons" />
      </div>
    </div>
  </div>
</template>

<script>
import { MDCRipple } from '@material/ripple'
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    outlined: {
      type: Boolean,
      default: false
    },
    fullBleedAction: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      slotObserver: undefined,
      mdcRipple: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-card--outlined': this.outlined
      }
    },
    contentClasses () {
      return {
        'mdc-card__primary-action': this.primaryAction
      }
    },
    actionClasses () {
      return {
        'mdc-card__actions--full-bleed': this.fullBleedAction
      }
    },
    primaryAction () {
      return this.$slots.actionableContent != null
    }
  },
  watch: {
    primaryAction (value) {
      value ? this.mdcRipple = MDCRipple.attachTo(this.$refs.content) : this.mdcRipple.destroy()
    }
  },
  mounted () {
    this.updateSlots()
    this.slotObserver = new MutationObserver(() => this.updateSlots())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })

    if (this.primaryAction) { this.mdcRipple = MDCRipple.attachTo(this.$refs.content) }
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
  },
  methods: {
    updateSlots () {
      if (this.$slots.fullBleedButton) {
        this.$slots.fullBleedButton.map((n) => {
          if (n.elm && n.elm.classList) {
            n.elm.classList.add('mdc-card__action')
            n.elm.classList.add('mdc-card__action--button')
          }
        })
      }
      if (this.$slots.actionButtons) {
        this.$slots.actionButtons.forEach((n) => {
          if (n.elm && n.elm.classList) {
            n.elm.classList.add('mdc-card__action')
            n.elm.classList.add('mdc-card__action--button')
          }
        })
      }
      if (this.$slots.actionIcons) {
        this.$slots.actionIcons.forEach((n) => {
          if (n.elm && n.elm.classList) {
            n.elm.classList.add('mdc-card__action')
            n.elm.classList.add('mdc-card__action--icon')
          }
        })
      }
    }
  }
}
</script>
