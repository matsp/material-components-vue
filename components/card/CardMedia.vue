<template>
  <div
    :class="classes"
    class="mdc-card__media"
  >
    <div
      v-if="$slots['default']"
      class="mdc-card__media-content"
    >
      <slot />
    </div>
  </div>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    square: {
      type: Boolean,
      default: true
    },
    sixteenNine: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes () {
      return {
        'mdc-card__media--square': this.square,
        'mdc-card__media--16-9': this.sixteenNine
      }
    }
  }
}
</script>
