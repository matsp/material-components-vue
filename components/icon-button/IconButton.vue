<template>
  <a
    v-if="href !== ''"
    :href="href"
    :data-toggle-on-content="toggleOnContent"
    :data-toggle-on-label="toggleOnLabel"
    :data-toggle-on-class="toggleOnClass"
    :data-toggle-off-content="toggleOffContent"
    :data-toggle-off-label="toggleOffLabel"
    :data-toggle-off-class="toggleOffClass"
    v-bind="$attrs"
    class="mdc-icon-button"
    :class="classes"
    v-on="$listeners"
    @MDCIconButtonToggle:change="$emit('change', $event.detail.isOn)"
  >
    <slot v-if="isIconButton" />
    <slot
      v-if="isToggleButtonViaSlots && !value"
      name="toggleOn"
    />
    <slot
      v-if="isToggleButtonViaSlots && value"
      name="toggleOff"
    />
  </a>
  <button
    v-else
    :data-toggle-on-content="toggleOnContent"
    :data-toggle-on-label="toggleOnLabel"
    :data-toggle-on-class="toggleOnClass"
    :data-toggle-off-content="toggleOffContent"
    :data-toggle-off-label="toggleOffLabel"
    :data-toggle-off-class="toggleOffClass"
    v-bind="$attrs"
    class="mdc-icon-button"
    :class="classes"
    v-on="$listeners"
    @MDCIconButtonToggle:change="$emit('change', $event.detail.isOn)"
  >
    <template v-if="this.icon && this.icon !== ''">
      {{ icon }}
    </template>
    <slot v-if="isIconButton" />
    <slot
      v-if="isToggleButtonViaSlots && !value"
      name="toggleOn"
    />
    <slot
      v-if="isToggleButtonViaSlots && value"
      name="toggleOff"
    />
  </button>
</template>

<script>
import { MDCRipple } from '@material/ripple'
import { MDCIconButtonToggle } from '@material/icon-button'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'value',
    event: 'change'
  },
  props: {
    toggleOnContent: {
      type: String,
      default: ''
    },
    toggleOnLabel: {
      type: String,
      default: ''
    },
    toggleOnClass: {
      type: String,
      default: ''
    },
    toggleOffContent: {
      type: String,
      default: ''
    },
    toggleOffLabel: {
      type: String,
      default: ''
    },
    toggleOffClass: {
      type: String,
      default: ''
    },
    value: {
      type: Boolean,
      default: false
    },
    href: {
      type: String,
      default: ''
    },
    icon: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      mdcIconButtonToggle: undefined,
      mdcRipple: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    isToggleButtonViaSlots () {
      return this.$slots.toggleOn && this.$slots.toggleOff && !this.isIconButton
    },
    isIconButton () {
      return this.$slots.default
    },
    isToggleButton () {
      return this.toggleOnContent !== '' && this.toggleOffContent !== ''
    },
    classes () {
      return {
        'material-icons': this.icon && this.icon !== ''
      }
    }
  },
  watch: {
    value (value) {
      if (typeof this.mdcIconButtonToggle !== 'undefined') {
        this.mdcIconButtonToggle.on = value
      }
    }
  },
  mounted () {
    this.update()
    this.slotObserver = new MutationObserver(() => this.update())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
  },
  beforeDestroy () {
    if (typeof this.mdcIconButtonToggle !== 'undefined') {
      this.mdcIconButtonToggle.destroy()
    }
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
    this.slotObserver.disconnect()
  },
  methods: {
    update () {
      if ((this.isToggleButton || this.isToggleButtonViaSlots) && typeof this.mdcIconButtonToggle === 'undefined') {
        this.mdcIconButtonToggle = MDCIconButtonToggle.attachTo(this.$el)
        this.mdcIconButtonToggle.on = this.value
      }
      if (this.isIconButton && typeof this.mdcRipple === 'undefined') {
        this.mdcRipple = MDCRipple.attachTo(this.$el)
        this.mdcRipple.unbounded = true
      }
    }
  }
}
</script>
