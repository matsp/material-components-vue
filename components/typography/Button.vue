<template>
  <span class="mdc-typography--button">
    <slot />
  </span>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin]
}
</script>
