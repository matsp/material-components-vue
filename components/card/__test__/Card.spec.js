import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import Card from '../Card.vue'
import CardMedia from '../CardMedia.vue'
import Button from '../../button/Button.vue'
import IconButton from '../../icon-button/IconButton.vue'

describe('Card Media', () => {
  it('should mount', () => {
    let wrapper = mount(CardMedia)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    let wrapper = mount(CardMedia)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-card__media')
  })

  it('should render as square', () => {
    let wrapper = mount(CardMedia, {
      propsData: {
        square: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-card__media--square')
  })

  it('should render as 16:9', () => {
    let wrapper = mount(CardMedia, {
      propsData: {
        sixteenNine: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-card__media--16-9')
  })
})

describe('Card', () => {
  it('should mount', () => {
    let wrapper = mount(Card)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.slotObserver).toBeDefined()
  })

  it('should render with no prop', () => {
    let wrapper = mount(Card)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-card')
  })

  it('should render as outlined', () => {
    let wrapper = mount(Card, {
      propsData: {
        outlined: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-card--outlined')
  })

  it('should render as primary action', () => {
    let wrapper = mount(Card, {
      slots: {
        actionableContent: 'content'
      }
    })
    expect(wrapper).toMatchSnapshot()
    let content = wrapper.find('.mdc-card__primary-action')
    expect(content).toBeDefined()
    expect(content.attributes().tabindex).toBe('0')
  })

  it('should render with full bleed action', () => {
    let wrapper = mount(Card, {
      propsData: {
        fullBleedAction: true
      },
      slots: {
        fullBleedButton: Button
      }
    })
    expect(wrapper).toMatchSnapshot()
    let actions = wrapper.find('.mdc-card__actions')
    expect(actions.classes()).toContain('mdc-card__actions--full-bleed')
    expect(actions.find('.mdc-button').classes()).toContain('mdc-card__action')
    expect(actions.find('.mdc-button').classes()).toContain('mdc-card__action--button')
  })

  it('should render with action buttons', () => {
    let wrapper = mount(Card, {
      slots: {
        actionButtons: [Button, Button]
      }
    })
    expect(wrapper).toMatchSnapshot()
    let actions = wrapper.find('.mdc-card__actions')
    actions.findAll('.mdc-button').wrappers.forEach(w => {
      expect(w.classes()).toContain('mdc-card__action')
      expect(w.classes()).toContain('mdc-card__action--button')
    })
  })

  it('should render with action icons', () => {
    let wrapper = mount(Card, {
      slots: {
        actionIcons: [IconButton, IconButton]
      }
    })
    expect(wrapper).toMatchSnapshot()
    let actions = wrapper.find('.mdc-card__actions')
    actions.findAll('.mdc-icon-button').wrappers.forEach(w => {
      expect(w.classes()).toContain('mdc-card__action')
      expect(w.classes()).toContain('mdc-card__action--icon')
    })
  })
})
