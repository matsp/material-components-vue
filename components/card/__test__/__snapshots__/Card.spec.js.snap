// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Media should render as 16:9 1`] = `
<div class="mdc-card__media mdc-card__media--square mdc-card__media--16-9">
  <!---->
</div>
`;

exports[`Card Media should render as square 1`] = `
<div class="mdc-card__media mdc-card__media--square">
  <!---->
</div>
`;

exports[`Card Media should render with no prop 1`] = `
<div class="mdc-card__media mdc-card__media--square">
  <!---->
</div>
`;

exports[`Card should render as outlined 1`] = `
<div class="mdc-card mdc-card--outlined">
  <!---->
</div>
`;

exports[`Card should render with action buttons 1`] = `
<div class="mdc-card">
  <div class="mdc-card__actions">
    <!---->
    <div class="mdc-card__action-buttons"><button class="mdc-button mdc-card__action mdc-card__action--button"> <span class="mdc-button__label"></span> </button><button class="mdc-button mdc-card__action mdc-card__action--button"> <span class="mdc-button__label"></span> </button></div>
    <!---->
  </div>
</div>
`;

exports[`Card should render with action icons 1`] = `
<div class="mdc-card">
  <div class="mdc-card__actions">
    <!---->
    <!---->
    <div class="mdc-card__action-icons"><button class="mdc-icon-button mdc-ripple-upgraded--unbounded mdc-card__action mdc-card__action--icon">
        <!---->
        <!----></button><button class="mdc-icon-button mdc-ripple-upgraded--unbounded mdc-card__action mdc-card__action--icon">
        <!---->
        <!----></button></div>
  </div>
</div>
`;

exports[`Card should render with full bleed action 1`] = `
<div class="mdc-card">
  <div class="mdc-card__actions mdc-card__actions--full-bleed"><button class="mdc-button mdc-card__action mdc-card__action--button"> <span class="mdc-button__label"></span> </button>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`Card should render with no prop 1`] = `
<div class="mdc-card">
  <!---->
</div>
`;

exports[`Card should render with primary action 1`] = `
<div class="mdc-card">
  <div tabindex="0" class="mdc-card__primary-action">
    <div class="mdc-card__media mdc-card__media--square">
      <!---->
    </div>
  </div>
  <!---->
</div>
`;

exports[`CardPrimaryAction should render with no prop 1`] = `<div tabindex="0" class="mdc-card__primary-action"></div>`;

exports[`CardPrimaryAction should render without ripple 1`] = `<div tabindex="0" class="mdc-card__primary-action"></div>`;
